mongo/search_index_view.go

```diff
@@ -0,0 +1,256 @@
+// Copyright (C) MongoDB, Inc. 2023-present.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License. You may obtain
+// a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+
+package mongo
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+
+	"go.mongodb.org/mongo-driver/bson"
+	"go.mongodb.org/mongo-driver/mongo/options"
+	"go.mongodb.org/mongo-driver/mongo/writeconcern"
+	"go.mongodb.org/mongo-driver/x/bsonx/bsoncore"
+	"go.mongodb.org/mongo-driver/x/mongo/driver"
+	"go.mongodb.org/mongo-driver/x/mongo/driver/operation"
+	"go.mongodb.org/mongo-driver/x/mongo/driver/session"
+)
+
+// SearchIndexView is a type that can be used to create, drop, list and update search indexes on a collection. A SearchIndexView for
+// a collection can be created by a call to Collection.SearchIndexes().
+type SearchIndexView struct {
+	coll *Collection
+}
+
+// SearchIndexModel represents a new search index to be created.
+type SearchIndexModel struct {
+	// A document describing the definition for the search index. It cannot be nil.
+	// See https://www.mongodb.com/docs/atlas/atlas-search/create-index/ for reference.
+	Definition interface{}
+
+	// The name for the index, if present.
+	Name *string
+}
+
+// List executes a listSearchIndexes command and returns a cursor over the search indexes in the collection.
+//
+// The aggregateOpts parameter is the aggregation options.
+//
+// The opts parameter can be used to specify options for this operation (see the options.ListSearchIndexesOptions
+// documentation).
+func (siv SearchIndexView) List(ctx context.Context, name *string,
```

Comment:
Is empty string a valid name? If not, we should treat empty string as "not specified" instead of using a string pointer. We should also note in the description that empty string matches all indexes.

Reply:
Any empty names passed to the creation functions will be returned as "default". However, the specs specify the name is optional but do not restrict it to be non-empty. For that reason, using a pointer gains better backward compatibility in my opinion.

Reply:
I'm a little worried about the usability problems that using string pointers introduces, especially because there is no Go syntax for declaring pointers to literal strings. We use that pattern in the `WriteConcern` type, but the negative usability impact seems greater than anticipated.

If we need to be able to distinguish between empty string and no value provided, would adding `name` to `ListSearchIndexesOptions` be a better option?

E.g. using string pointer:
```go
coll.SearchIndexes().List(ctx, nil)

name := "myIndex"
coll.SearchIndexes().List(ctx, &name)
```

E.g. using `ListSearchIndexOptions`:
```go
coll.SearchIndexes().List(ctx)

coll.SearchIndexes().List(ctx, options.ListSearchIndexes().Name("myIndex"))
```

Another consideration is that the `$listSearchIndexes` aggregation pipeline supports [both "id" and "name filters](https://www.mongodb.com/docs/manual/reference/operator/aggregation/listSearchIndexes/#command-fields). If we ever need to support the "id" filter in the future, we would need to add it somewhere. We could accept a `filter interface{}` parameter, similar to [ListCollections](https://pkg.go.dev/go.mongodb.org/mongo-driver/mongo#Database.ListCollections). However, that would be less usable for a simple string filter because users would have to figure out how to assemble the filter document themselves. Also, the filter document format for `ListCollections` is very different than the filter document format for `$listSearchIndexes`.

E.g. using filter:
```
coll.SearchIndexes().List(ctx, nil)
coll.SearchIndexes().List(ctx, bson.D{{"name", "myIndex"}})
```

I don't think any of those are the obvious best choice. Which do you prefer?

Reply:
OK, I've introduced `SearchIndexesOptions` with a `SetName(string)` method to align with the interfaces such as `AggregateOptions`.

Maybe we can provide an `Optional` type with the help of generic functions in the 2.0 driver. I don't know if it is a convention in the Golang.